src/server/conn_context.cc

```diff
@@ -165,42 +165,63 @@ void ConnectionContext::ChangePSub(bool to_add, bool to_reply, CmdArgList args)
 
   if (to_reply) {
     const char* action[2] = {"punsubscribe", "psubscribe"};
+    if (result.size() == 0) {
+      return SendSubscriptionChangedResponse(action[to_add], std::nullopt, 0);
+    }
 
     for (size_t i = 0; i < result.size(); ++i) {
-      (*this)->StartArray(3);
-      (*this)->SendBulkString(action[to_add]);
-      (*this)->SendBulkString(ArgS(args, i));
-      (*this)->SendLong(result[i]);
+      SendSubscriptionChangedResponse(action[to_add], ArgS(args, i), result[i]);
     }
   }
 }
+void ConnectionContext::UnsubscribeAll(bool to_reply) {
+  if (to_reply && (!conn_state.subscribe_info ||
+                   conn_state.subscribe_info->channels.empty())) {
+    return SendSubscriptionChangedResponse("unsubscribe", std::nullopt, 0);
+  }
+  StringVec channels(conn_state.subscribe_info->channels.begin(),
+                     conn_state.subscribe_info->channels.end());
+  CmdArgVec arg_vec(channels.begin(), channels.end());
+
+  ChangeSubscription(false, to_reply, CmdArgList{arg_vec});
+
+}
+
+void ConnectionContext::PUnsubscribeAll(bool to_reply) {
+  if (to_reply && (!conn_state.subscribe_info ||
+                   conn_state.subscribe_info->patterns.empty())) {
+    return SendSubscriptionChangedResponse("punsubscribe", std::nullopt, 0);
+  }
 
+  StringVec patterns(conn_state.subscribe_info->patterns.begin(),
+                     conn_state.subscribe_info->patterns.end());
+  CmdArgVec arg_vec(patterns.begin(), patterns.end());
+  ChangePSub(false, to_reply, CmdArgList{arg_vec});
+
+}
+
+void ConnectionContext::SendSubscriptionChangedResponse(
+    string_view action, std::optional<string_view> topic, unsigned count) {
+  (*this)->StartArray(3);
+  (*this)->SendBulkString(action);
+  topic.has_value() ? (*this)->SendBulkString(topic.value())
```

Comment:
I prefer avoid using operator `? :`  for action-only conditions.

Reply:
:+1: As I understood it you prefer this: af4272a71d380c036887ad473ab9ecd8a093def2?

Reply:
yes